velox/substrait/SubstraitToVeloxPlan.cpp

```diff
@@ -282,49 +342,115 @@ std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
   }
   auto outputType = ROW(std::move(outNames), std::move(veloxTypeList));
 
-  auto tableScanNode = std::make_shared<core::TableScanNode>(
-      nextPlanNodeId(), outputType, tableHandle, assignments);
-  return tableScanNode;
+  if (sRead.has_virtual_table()) {
+    return toVeloxPlan(sRead, pool, outputType);
+
+  } else {
+    auto tableScanNode = std::make_shared<core::TableScanNode>(
+        nextPlanNodeId(), outputType, tableHandle, assignments);
+    return tableScanNode;
+  }
+}
+
+std::shared_ptr<const core::PlanNode> SubstraitVeloxPlanConverter::toVeloxPlan(
+    const ::substrait::ReadRel& readRel,
+    facebook::velox::memory::MemoryPool* pool,
+    std::shared_ptr<const RowType> type) {
+  ::substrait::ReadRel_VirtualTable readVirtualTable = readRel.virtual_table();
+  int64_t numRows = readVirtualTable.values_size();
+  int64_t numColumns = type->size();
+  int64_t valueFieldNums = readVirtualTable.values(numRows - 1).fields_size();
+  std::vector<RowVectorPtr> vectors;
+  vectors.reserve(numRows);
+
+  int64_t batchSize = valueFieldNums / numColumns;
+
+  for (int64_t row = 0; row < numRows; ++row) {
+    std::vector<VectorPtr> children;
+    ::substrait::Expression_Literal_Struct rowValue =
+        readRel.virtual_table().values(row);
+    auto fieldSize = rowValue.fields_size();
+    VELOX_CHECK_EQ(fieldSize, batchSize * numColumns);
+
+    const auto& childrenSize = type->children().size();
+    for (int64_t col = 0; col < childrenSize; ++col) {
+      const TypePtr& outputChildType = type->childAt(col);
+      std::vector<variant> batchChild;
+      batchChild.reserve(batchSize);
+      for (int64_t batchId = 0; batchId < batchSize; batchId++) {
+        // each value in the batch
+        auto fieldIdx = col * batchSize + batchId;
+        ::substrait::Expression_Literal field = rowValue.fields(fieldIdx);
+
+        auto expr = exprConverter_->toVeloxExpr(field);
+        if (auto constantExpr =
+                std::dynamic_pointer_cast<const core::ConstantTypedExpr>(
+                    expr)) {
+          if (!constantExpr->hasValueVector()) {
+            batchChild.emplace_back(constantExpr->value());
+          } else {
+            VELOX_UNSUPPORTED(
+                "Values node with complex type values is not supported yet");
+          }
+        } else {
+          VELOX_FAIL("Expected constant expression");
+        }
+      }
+      children.emplace_back(
+          setCellFromVariant(outputChildType, batchChild, pool));
+    }
+
+    // make rowVectors
```

Comment:
Comments should be full sentences, e.g. start with a capital letter and end with a period.